---
title: "NAFLD experiments (BCN-Mar Coll), CellRanger v4 - Data import"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, message=FALSE}
rm(list=ls())
library(Seurat)
library(ggplot2)

library(readxl)
library(knitr)
library(data.table)
library(cowplot)

data_storage_folder = file.path("./Results/data_storage","2021-05-31")
if (!file.exists(data_storage_folder)) dir.create(data_storage_folder, recursive=T)


```

# Overview

We here use data from control experiments of BL6 mice fed with different diets, from liver or BM, with or without CCl4 treatment and from different timepoints. Myeloid leukocytes were extracte by FACS sorting and measured on a 10x Chromium system (8 channel microfluidics chip). 
Gene expression read counts per cell were generated from sequence data using CellRanger v4. 

We here use strict filtering in particular for cells with high Mt RNA proportions. 

## Sample description

```{r}
ed = as.data.frame(read_excel("../SampleDescription.xlsx", sheet = 1), stringsAsFactors = F)
rownames(ed) <- NULL
kable(ed)
```

# Data import

```{r}
reload_data_stage1 = F

stage1_file = file.path(data_storage_folder, "All_Raw_scData_stage1.Rdata")

if(reload_data_stage1) {
  mat_files = list.files(path="../CellRanger/", pattern = "*.mtx.gz", recursive = T, full.names = T)
  mat_files = mat_files[grepl("filtered_feature_bc_matrix",mat_files)]
  folders = dirname(mat_files)
  sample_ids = unlist(sapply(strsplit(folders,"/"), function(x) x[length(x)-2] ) )
  names(folders) = sample_ids
  
  all_sample_data = list()
  for (n in sample_ids) {
    dir(path=folders[n])
    tmp = Read10X( folders[n] )
    all_sample_data[[n]] = CreateSeuratObject(tmp, project=n)
  }
  
  all_ge_data_raw = merge(all_sample_data[[1]], all_sample_data[2:length(all_sample_data)])
  rm(all_sample_data)
  gc(verbose=F)
  
  rownames(ed) = ed$Barcode
  
  ed$sample_id_final = ed$Barcode
  
  meta = as.data.table(all_ge_data_raw@meta.data)
  meta[, orig_cell_id:= rownames(all_ge_data_raw@meta.data)]
  meta[, subfolder:=ed[orig.ident,"FILES"] ]
  meta[, sample_id_final:=orig.ident]
  
  meta[, group:=ed[sample_id_final, "Condition"]]
  meta[, fibrotic_stage:= ed[sample_id_final, "Fibrotic stage"]] 
  meta[, label:= ed[sample_id_final, "Fibrotic stage"]]
  meta[, sample_label:=ed[sample_id_final, "NAME"]]
  meta[, orig_sample_id:=ed[sample_id_final, "SAMPLE ID"]]
  
  for (cc in c("orig_cell_id", "sample_id_final", "group","label", "sample_label","fibrotic_stage")) {
    all_ge_data_raw = AddMetaData(all_ge_data_raw, meta[[cc]], col.name=cc)
  }
  
  save(all_ge_data_raw, ed,  file=stage1_file)
} else {
  load(stage1_file)
}


```

## Cells per sample (unfiltered)

```{r}
meta = all_ge_data_raw@meta.data
table(meta$sample_label)
table(meta$label)
table(meta$group)
```

# Quality control

## Detected genes

```{r, fig.width = 8}
ggplot(meta) + geom_boxplot(aes(x=orig.ident, y=nFeature_RNA, fill=group)) + theme(axis.text.x = element_text(angle=45, hjust=1)) + ylab("Number of detected genes") + xlab("Sample") + scale_y_log10()
```

## UMIs

```{r, fig.width = 8}
ggplot(meta) + geom_boxplot(aes(x=orig.ident, y=nCount_RNA, fill=group)) + theme(axis.text.x = element_text(angle=45, hjust=1)) + ylab("Total RNA counts") + xlab("Sample") + scale_y_log10()
```


## Overview

```{r}
all_ge_data_raw <- PercentageFeatureSet(all_ge_data_raw, "^MT-|^mt-", col.name = "percent_mito")
all_ge_data_raw <- PercentageFeatureSet(all_ge_data_raw, "^RP[SL]|^Rp[sl]", col.name = "percent_ribo")
```

### General features

```{r, fig.width=14, fig.height=10}
feats <- c("nFeature_RNA","nCount_RNA","percent_mito","percent_ribo")
VlnPlot(all_ge_data_raw, group.by= "sample_label", features = feats, pt.size = 0.01,ncol = 2) + NoLegend()
```

### Pairwise parameter correlations

```{r, fig.width=14, fig.height=12}
cowplot::plot_grid(ncol = 2, nrow = 2, 
  FeatureScatter(all_ge_data_raw, "nCount_RNA"  , "nFeature_RNA", group.by = "sample_label", pt.size = .5) + guides(colour = guide_legend(override.aes = list(size=4))),
  FeatureScatter(all_ge_data_raw, "percent_mito", "nFeature_RNA", group.by = "sample_label", pt.size = .5) + guides(colour = guide_legend(override.aes = list(size=4))),
  FeatureScatter(all_ge_data_raw, "percent_ribo", "nFeature_RNA", group.by = "sample_label", pt.size = .5) + guides(colour = guide_legend(override.aes = list(size=4))),
  FeatureScatter(all_ge_data_raw, "percent_ribo", "percent_mito", group.by = "sample_label", pt.size = .5) + guides(colour = guide_legend(override.aes = list(size=4)))
)

```

### Pairwise feature distribution densities

```{r, fig.width=14, fig.height=10}
meta = as.data.table(all_ge_data_raw@meta.data)

ggplot(meta) + geom_density2d_filled(aes(x=nCount_RNA, y=nFeature_RNA), contour_var = "ndensity" ) + facet_wrap(~sample_label) + ggtitle("nCount_RNA")
ggplot(meta) + geom_density2d_filled(aes(x=percent_ribo, y=nFeature_RNA), contour_var = "ndensity" ) + facet_wrap(~sample_label) + ggtitle("percent ribo")
ggplot(meta) + geom_density2d_filled(aes(x=percent_mito, y=nFeature_RNA), contour_var = "ndensity" ) + facet_wrap(~sample_label) + ggtitle("percent mitochondrial")
```

### Thresholds for numbers of detected genes and percentage of mitochondrial genes

```{r}
hist(meta$nFeature_RNA, 1000, main="Number of detected genes per cell"); abline(v=c(200,3500), col="red")
hist(meta$percent_mito, 1000, main="Percent mitochondrial RNA"); abline(v=20, col="red")
```


```{r}
ggplot(meta, aes(x=nFeature_RNA)) + geom_density() + facet_wrap(~sample_label) + geom_vline(xintercept=c(300,5000)) + ggtitle("Number of detected features")
ggplot(meta, aes(x=percent_mito)) + geom_density() + facet_wrap(~sample_label) + geom_vline(xintercept=c(20)) + ggtitle("Percent mitochondrial transcripts")
```


# Filtering 

We exclude cells with less than 300 detected genes, those with > 5000 detected genes (possible doublets) and those with clear signs of cellular stress (percentage of mitochondrial RNA > 20%). 

```{r}
feature_low = 300
feature_high = 5000
max_mito = 20

all_ge_data_raw = AddMetaData(all_ge_data_raw, (all_ge_data_raw$nFeature_RNA > feature_low) & (all_ge_data_raw$nFeature_RNA < feature_high) & (all_ge_data_raw$percent_mito < max_mito), col.name="CellFilterKeep")

ribo.genes <- grepl(pattern = "^RP[SL]|^Rp[sl]", x = rownames(x = all_ge_data_raw@assays$RNA))
mito.genes <- grepl(pattern = "^MT-|^mt-", x = rownames(x = all_ge_data_raw@assays$RNA))

selected_c <- WhichCells(all_ge_data_raw, expression = CellFilterKeep )
selected_f <- rownames(all_ge_data_raw)[ Matrix::rowSums(all_ge_data_raw) >= 2 & !(ribo.genes | mito.genes)]
```


## Number of removed/kept cells per condition

```{r}
m = table(all_ge_data_raw$sample_label, ifelse(all_ge_data_raw$CellFilterKeep, "Keep","Remove") )
m
```

```{r}
barplot(t(m), col=c("blue","red"), main="Filtering results")
legend("topright", legend=c("Kept","Removed"), fill = c("blue","red"))
```


```{r}
all_ge_data_filtered <- subset(all_ge_data_raw, features=selected_f, cells=selected_c)

dim(all_ge_data_filtered)
```

Raw and filtered data is saved to individual files *All_Raw_scData.Rdata* and *All_Raw_scData.Rdata*.

```{r}
save(all_ge_data_raw, ed,  file=file.path(data_storage_folder, "All_Raw_scData.Rdata"))
save(all_ge_data_filtered, ed, file=file.path(data_storage_folder, "All_Filtered_scData.Rdata"))
```


# Software versions

```{r}
sessionInfo()
```

